### PR TITLE
Remove ApplyNoError from Schema

### PR DESCRIPTION
**Pull request by andrewkroh**
The ApplyNoError function is not needed because users can just call Apply and ignore the error.

See #3807

### CODE REVIEW

**Comment by andrewkroh** [comment]
jenkins, test it

File: metricbeat/schema/schema.go [hunk 1]
```diff
@@ -71,19 +71,13 @@ func (o Object) HasKey(key string) bool {
 
 // ApplyTo adds the fields extracted from data, converted using the schema, to the
 // event map.
-func (s Schema) ApplyTo(event common.MapStr, data map[string]interface{}) (common.MapStr, *Errors) {
```
Comment (ruflin) [inline metricbeat/schema/schema.go]: Why did you change *Errors to error? The idea is that after `Apply` the `Errors` is available and checks can be made if required or invalid fields were missing to create a more detailed error report.

Comment (andrewkroh) [inline metricbeat/schema/schema.go]: I would strongly prefer to use `error` in public interfaces. Using an specialized error type forces a strong coupling with the caller and makes the API more brittle. Callers that care about the precise error details can use a type switch or a type assertion to look for specific errors and extract details. Or for really common checks, the package can expose helper fucntions to inspect the error (e.g. `os.IsNotExists`, `os.IsPermission`).

This how most of the stdlib works and the methods that have specialized error types document what they return for callers that need the details (e.g. `os.Open` says "If there is an error, it will be of type *PathError."). 

This is a great [blog post](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) about the different error techniques.

Comment (ruflin) [inline metricbeat/schema/schema.go]: Type switches are mostly going to solve the problem or helper functions. One problem of the implementation could be that it is called error. Perhaps it should be renamed. In case there are some "optional" errors, error is not nil but should be to support the default error handling `if err != nil`. The information returned is more some meta information about the event. 

I understand the reasoning behind the public interface should return error but one advantage of having `*Errors` returned is that it forces the person implementing the code, to deal with it and understand the logic of it. The person implementing it should make a decision on how to deal with the different types of "Errors". As I can see us adding more logic to it in the future we could rename it to "Meta" or something like that and this object has a method `GetError`. So it would not be directly and error. Just some alternative ideas on how we could deal with it.

Thanks for the blog post. One interesting method I was not aware of is `errors.Wrap` that we should probably use more often in our code.

Comment (andrewkroh) [inline metricbeat/schema/schema.go]: > advantage of having *Errors returned is that it forces the person implementing the code, to deal with it and understand the logic of it

I agree that the caller needs to deal with the error, but I don't think we should force them to understand the error. Returning a non-nil `error` should indicate failure only (so let's separate errors from warnings). If the caller wishes to inspect the error then there is a way to look at the individual errors caused by absent required fields.

Having a meta object to indicate the missing optional fields sounds like a good option to me. Basically we would separate errors from warnings. And warnings could be a separate return argument and distinct type. Outside of tests, I would expect most callers to be ignoring this meta return value with the list of absent optional fields (maybe with the exception of logging?).

If you agree, let's put this in the queue for #3807 and work it in a separate PR.

Comment (ruflin) [inline metricbeat/schema/schema.go]: Agree on the above. But that requires changes to the code as right now also missing optional fields return an error (and not nil)
